[v2] Web-based help

kdaily opened the pull request:
### Description of changes


Add in a configuration option that can open CLI help pages in the browser or print the URL to the remote CLI reference guide.

The new configuration option is available in the shared configuration file as `cli_help_output`. It has four valid values:

1. Unset: default behavior opening the man page in the terminal.
1. `terminal`: default behavior opening the man page in the terminal.
1. `browser`: open the man page as HTML in the default browser.
1. `url`: print the URL to the remote AWS CLI reference guide for the version of the AWS CLI installed.

When set to `browser`, a warning message is printed to the terminal notifying the user that the default web browser is being opened, and an error message in the case the browser cannot be opened.

When set to `url`, the value for client side paging is respected (e.g., controlled through `AWS_PAGER` environment variable).

###  Implementation notes

This refactors the existing help renderer to move the previous parent class `PagingHelpRenderer` to a sub-class of a `HelpRenderer`. It refactors to potentially send output to different destinations, currently a pager or a browser.

Then, this adds a new renderer base class called the `BrowserHelpRenderer` that swaps out a private method that sends rendered output to the pager with one that sends it to the browser. The rendered content is written to a temporary file and then the users default browser is opened. Separate OS-specific renders inherit, as the paging-based renderers did. An OS-specific sub-class is added that uses this behavior (`PosixBrowserHelpRenderer`) to render the output using `groff` or `mandoc` using `html` instead of `ascii` output, and a `WindowsBrowserHelpRenderer` swaps out the text writer for an HTML writer.

Help command classes (derived from `HelpCommand`) now have a `url` property. Additionally, the `HelpCommand` now uses its own botocore session to determine if its help should be the documentation contents or the URL.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

aemous commented:
> When set to url, the value for client side paging is respected (e.g., controlled through `AWS_PAGER` environment variable).

Is there a case where this would print using the pager? I would think since it's just printing a URL it would just be a single line. Is this for the case that the URL is sufficiently long that the client would benefit from a paging program?

aemous reviewed (changes requested):
Looks good overall. Had some small nits and questions.

Would also like to see some tests added as you already mentioned in the PR description.

Would be nice to see some example output for each config value in the PR description for a single service/operation.

kdaily commented on awscli/help.py:
```diff
@@ -88,54 +138,62 @@ def get_pager_cmdline(self):
+        html_file.close()
 
+        try:
+            print("Opening help file in the default browser.")
```
I think here we want this to be explictly shown to the user on stdout. This would be similar to this message on SSO login:

https://github.com/aws/aws-cli/blob/86520617901307c5078fe74712ea61ebf5f38310/awscli/customizations/sso/utils.py#L199

kdaily replied:
Looks like `uni_print` is specifically for customizations. We use the std lib `print` in autocomplete as well:

https://github.com/aws/aws-cli/blob/9c8f131b89c2f13c87132446c93c21b467d9dafe/awscli/autocomplete/main.py#L62

kdaily commented on awscli/help.py:
```diff
@@ -88,54 +138,62 @@ def get_pager_cmdline(self):
+        # This is a temporary fix to allow the manpages for these commands
+        # to be rendered. Long term, we should avoid enumerating over all
+        # enums inline for the JSON syntax snippets.
+        'line_length_limit': 50_000
```
Good question - this is actually taken from here though, its pre-existing:

https://github.com/aws/aws-cli/blob/86520617901307c5078fe74712ea61ebf5f38310/awscli/help.py#L72-L81

kdaily commented:
> > When set to url, the value for client side paging is respected (e.g., controlled through `AWS_PAGER` environment variable).
> 
> Is there a case where this would print using the pager? I would think since it's just printing a URL it would just be a single line. Is this for the case that the URL is sufficiently long that the client would benefit from a paging program?

I erred on the side that when we print the result of a command to the console, we always pass it through the pager, to maintain consistency.

aemous commented:
> I added the ability to set the help output format via an environment variable. This provides users a way to set it per invocation without adding a global flag, which is tricky to add since it should only apply to `aws * help` commands.

1. Have we considered just adding an argument to the `aws help` command rather than global? 
2. (non-blocking) I find it a bit weird that we are adding an env/config var that only modifies the behavior of a single command. But this isn't the first env var to do so, `AWS_CLI_S3_MV_VALIDATE_SAME_S3_PATHS` is an env var that only modifies `aws s3 mv`. This is just a note/consideration, not a requested change.

aemous reviewed (changes requested):
Looks like some tests are still failing, may be flaky based on the OS/runtime versions.

Requested some (mostly minor) changes and left some questions

kdaily commented on awscli/help.py:
```diff
@@ -88,59 +139,69 @@ def get_pager_cmdline(self):
+        self._send_output_to_browser(output)
+
+    def _send_output_to_browser(self, output):
+        html_file = tempfile.NamedTemporaryFile(
+            "wb", suffix=".html", delete=False
+        )
+        html_file.write(output)
+        html_file.close()
```
Noted as an improvement to determine a caching strategy.

kdaily commented on awscli/customizations/commands.py:
```diff
@@ -408,7 +415,12 @@ def __call__(self, args, parsed_globals):
         docevents.generate_events(self.session, self)
-        self.renderer.render(self.doc.getvalue())
+
+        if self._help_output_format == 'url':
+            self.renderer.render(self.url.encode())
+        else:
+            self.renderer.render(self.doc.getvalue())
+
```
I looked at the parameters that control the autoprompt mode. It follows a similar pattern that falls back to a default if an unknown value is used. I tested this by setting an unknown value in the config file like this:

```
[default]
cli_auto_prompt = absolutely
```

The only valid values for `cli_auto_prompt` are `off`, `on`, and `on-partial`. This code falls back to `off` if any other value is given.

https://github.com/aws/aws-cli/blob/a8d8f06152c3c4b8a5726e73c01fb30a9e3b2975/awscli/clidriver.py#L224-L236

aemous replied:
Alright thanks for confirming that.

kdaily commented on awscli/customizations/commands.py:
```diff
@@ -383,6 +384,12 @@ def examples(self):
 
+    @property
+    def url(self):
+        if len(self.command_table) > 0:
```
Customizations, including custom commands and things like wizards, don't have a separate help command interface, they inherit from `BasicCommand`. This logic is taken from the docs generator to differentiate between a command (which indexes sub-commands) and sub-commands that are the leaves of documentation pages.

kdaily commented on awscli/help.py:
```diff
@@ -68,6 +91,34 @@ class PagingHelpRenderer:
+
+
+class PagingHelpRenderer(HelpRenderer):
+    """Interface for a help renderer.
```
I think its fine to leave it. Even this module is inconsistent (see `ProviderHelpCommand` or `OperationHelpCommand`, for example). The PEP for docstrings suggests first line is in line with the opening triple quote.

kdaily commented:
> > I added the ability to set the help output format via an environment variable. This provides users a way to set it per invocation without adding a global flag, which is tricky to add since it should only apply to `aws * help` commands.

For the time being until an interface is agreed upon, I removed the environment variable option. 

>     1. Have we considered just adding an argument to the `aws help` command rather than global?

To add to `help` only, it would need a different interface. It doesn't look like `arg_table` is used for help commands at the moment. An offline suggestion to add a new help subcommand (like `aws help browser`), which would be alongside `aws help topics`, is also possible. However, this would need to be available on every sub-command too (so `aws sts get-caller-identity help browser`). It would be possible to add a DIFFERENT sub-command, like `aws help-browser`), but this is adding yet another help system, and it was deemed undesirable.
 
>     2. (non-blocking) I find it a bit weird that we are adding an env/config var that only modifies the behavior of a single command. But this isn't the first env var to do so, `AWS_CLI_S3_MV_VALIDATE_SAME_S3_PATHS` is an env var that only modifies `aws s3 mv`, so there is precedent. This is just a note/consideration, not a requested change.

I'm not sure that I understand this concern? Setting the config variable (or the since removed env variable) would change the behavior for every help command provided in the AWS CLI; though maybe what you mean is it ONLY applies when adding the sub-command `help`, and yes this is indeed novel.

aemous reviewed (approved):
LGTM.

aemous reviewed (approved):
LGTM

kdaily commented:
Merging with a failed but unrelated prompttoolkit test.
